### ISSUE TITLE
cabal can’t build package if there is no build plan for an unrelated package

### ISSUE THREAD

**Issue by maralorn** [haskell/cabal#9291]
**Describe the bug**

`cabal build package:foo` seems to run the solver for all packages mentioned in `cabal.project`, even if they are completely unnecessary to build package foo.

**To Reproduce**

Preparation:
```
$ mkdir test-project
$ cd test-project
$ echo "packages: */*.cabal" > cabal.project
$ mkdir a b
$ (cd a; cabal init)
$ (cd b; cabal init)
```
Now `cabal build b` should work fine.

But introduce an unknown dependency. e.g. `fuiae` in `a/a.cabal`.
Run `cabal build b` again:

```
  ~/test-project ❯ cabal build package:b                                                                                15s   02:38:45
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: a-0.1.0.0 (user goal)
[__1] unknown package: fuiae (dependency of a)
[__1] fail (backjumping, conflict set: a, fuiae)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: a, fuiae
```

**Expected behavior**
cabal should be able to build a package no matter whether other packages have a valid build plan.

**System information**
 - Operating system: nixos-unstable
 - The Glorious Glasgow Haskell Compilation System, version 9.4.6
 - cabal-install version 3.10.1.0
 - compiled using version 3.10.1.0 of the Cabal library 

**Additional context**
I encountered this in a larger project, this is just a minimal reproducer.

**Comment by Mikolaj** [comment]
This is related to a similar question for components of a single package: https://github.com/haskell/cabal/issues/7883. IIRC we decided to build a plan for all components and compile only the requested ones, unless the user specifically requests a more modest plan. It's not implemented yet.

Packages of a single project is a different question and the compilation is more costly, but it would be good to agree on principles at least.

**Comment by chreekat** [comment]
@Mikolaj does #7883 actually specify the proposal somewhere? It's a long thread, and suggestions, commentary, and analysis are freely intermixed...

Does Cabal have a proposal process? :smile:

**Comment by Mikolaj** [comment]
> Does Cabal have a proposal process? 😄

Yes. :D A ticket is opened, discussed in various ways and when the pr-welcome label is assigned, the proposal is considered accepted (though the resulting PR may not be, e.g., if it uncovers a flaw in the design).

> I agree that there is a relevant difference between components and packages here.

Right. So we may, e.g., have different defaults --- for components we'd create build plan for all, but for packages, only for the explicitly requested ones. However, both default should be overridable IMHO. So there at at least 2 separate questions: defaults and how to override them.

> So for hls to work, I currently need to edit cabal.project.

Yes, editing `cabal.project` proves something is very wrong here. Let's brainstorm. Would permitting `cabal.project.local` to override package list be strong enough? Convenient enough? Is such a functionality bad for any other reasons? Is a commandline flag better? What are we missing? Any out of the box approaches?

**Comment by maralorn** [comment]
A `cabal.project` (or `cabal.project.local`) option `plan-unrequired-packages` with default false would be a solution.

Don‘t know if it’s a nice one.^^

**Comment by maralorn** [comment]
Out of the scope of this issue but closely related would be to have the ability to force using the package-db or hackage for packages which we depend on and are mentioned in cabal.project. That way one can e.g. substitute a big package in the repo from CI via Nix while one is only working on another package in the project.

It’s closely related because it could also be solved by overriding `packages:` in `cabal.project.local`.

**Comment by maralorn** [comment]
> A `cabal.project` (or `cabal.project.local`) option `plan-unrequired-packages` with default false would be a solution.

Or maybe we can just downgrade a solver error (and maybe even syntax error) for an unrequired package to a warning?

**Comment by mpickering** [comment]
I think it makes sense to solve for all the packages in a `cabal.project`. 

Wouldn't it also be confusing if you could `cabal build exe:a` and `cabal build exe:b` in a project, but not `cabal build exe:a exe:b`? 

It would also be very confusing if the build plans for `exe:a and `exe:b` were constructed separately (you could end up building the same library against different dependencies.

**Comment by maralorn** [comment]
I guess I would be okay with `cabal build exe:a` and `cabal build exe:a exe:b` giving different build plans for `exe:a`, but I see the trade-off now. I thought the solver was simply being called multiple times, which in retrospect doesn‘t make much sense.
Then maybe we do `plan-unrequired-packages` but default it to true?

**Comment by gbaz** [comment]
>  Currently, I workaround this issue by always adapting my cabal.project to the packages I am currently working on. But cabal.project is a checked in file in the git repo and it feels weird to change it back and forth.

I would suggest using the new cabal project includes feature. You can have a core project file, and then additional files that import it, and pick between them using the --project-file flag on the command line.

**Comment by maralorn** [comment]
Picking on the command line is sadly not an option with hls.

**Comment by andreabedini** [comment]
I am not sure I understand the use case here. If the two packages `a` and `b` in the reproducible example are part of the same project they are meant to be built together. This means we cannot know the dependencies of `b` without also knowing those of `a` (what if a transitive dependency of `a` makes `b` unbuildable?).

That all packages in a project are planned together (with some caveats) is a feature not a bug.

**Comment by chreekat** [comment]
> @chreekat I think the conclusion was this: https://github.com/haskell/cabal/issues/7883#issuecomment-1035337279

Ah, well, that might be a good reference table, but I guess I was hoping for a clear statement "Cabal always [never?] builds [all? some? none?] components, unless [flags? cabal.project options?]". The table doesn't really tell me what the intent was.

**Comment by Mikolaj** [comment]
> "Cabal always [never?] builds [all? some? none?] components, unless [flags? cabal.project options?]"

I don't know the whole picture, but at least for tests, cabal-install at the moment does the wrong thing, which is not planning for all components (including the test component), so the default is wrong, and also the ability to configure this is not perfect. The table in https://github.com/haskell/cabal/issues/7883#issuecomment-1035337279 addresses both problems for the tests case.

I agree it would be good to determine what the current situation is generally with defaults for plans of components and packages and then to think what defaults and what configuration capability we'd need. I think the latter, configuration capability, would be less controversial, but not necessarily trivial, as the table demonstrates.

**Comment by gbaz** [comment]
> Picking on the command line is sadly not an option with hls.

hls lets you explicitly pick the cabalProject file, which should let the proposed solution work https://github.com/haskell/hie-bios#cabal

**Comment by maralorn** [comment]
I am sorry. I was fooled by the official documentation saying "The complete configuration is a subset of" and then not listing the "cabalProject" option.

https://haskell-language-server.readthedocs.io/en/latest/configuration.html#custom-program

**Comment by gbaz** [comment]
No worries, I had to hunt to find it in the docs and would absolutely encourage you to make a PR to somewhere around where you highlighted to make it more evident.

**Comment by andreabedini** [comment]
> hls lets you explicitly pick the cabalProject file, which should let the proposed solution work https://github.com/haskell/hie-bios#cabal

TIL!

**Comment by sebright** [comment]
Closing, since it sounds like the issue can be solved by using different project files, with project file includes.
